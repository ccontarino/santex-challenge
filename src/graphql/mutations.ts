import { gql } from '@apollo/client';

export const ADD_ITEM_TO_ORDER = gql`
  mutation addItemToOrder($productVariantId: ID!, $quantity: Int!) {
    addItemToOrder(productVariantId: $productVariantId, quantity: $quantity) {
      __typename
      ...on Order {
        id
        createdAt
        state
        total
        subTotal
        totalWithTax
        currencyCode
      }
    }
  }
`;
