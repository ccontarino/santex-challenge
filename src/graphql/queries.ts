import { gql } from '@apollo/client';

export const GET_PRODUCTS = gql`
  query {
    products {
      totalItems
      items {
        id
        name
        description
        variants {
          id
          price
          productId
          name
          assets {
            preview
            source
          }
        }
        assets {
          source
          preview
        }
      }
    }
  }
`;

export const GET_ORDER_BY_ID = gql`
  query getOrder($id: ID!) {
    order(id: $id) {
      id
      total
      subTotal
      totalWithTax
    }
  }
`;
