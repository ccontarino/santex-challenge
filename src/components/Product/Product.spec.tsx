import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import ProductComponent from './Product';
import { ProductVariant } from '../../types/types';
import { ADD_ITEM_TO_ORDER } from '../../graphql/mutations';

const mockVariants: ProductVariant[] = [
  {
    id: 'Variant 1',
    name: 'Variant 1',
    price: 100,
  },
  {
    id: 'Variant 2',
    name: 'Variant 2',
    price: 200,
  },
];

describe('ProductComponent', () => {
  it('renders product title, image, description, variant selector and price', () => {
    render(
      <MockedProvider>
        <ProductComponent
          id="1"
          productTitle="Product Title"
          productImage="https://example.com/image.jpg"
          productDescription="Product Description"
          variants={mockVariants}
        />
      </MockedProvider>
    );

    expect(screen.getByText('Product Title')).toBeInTheDocument();

    expect(screen.getByText('Product Description')).toBeInTheDocument();
    expect(screen.getByText('Variant 2')).toBeInTheDocument();
    expect(screen.getByText('Buy $100')).toBeInTheDocument();
  });

  it('selects a variant when clicked', () => {
    render(
      <MockedProvider>
        <ProductComponent
          id="1"
          productTitle="Product Title"
          productImage="https://example.com/image.jpg"
          productDescription="Product Description"
          variants={mockVariants}
        />
      </MockedProvider>
    );

    fireEvent.click(screen.getByText('Variant 2'));

    waitFor(
      () => {
        expect(screen.getByText('Buy $200')).toBeInTheDocument();
      },
      { timeout: 1000 }
    );
  });
});
