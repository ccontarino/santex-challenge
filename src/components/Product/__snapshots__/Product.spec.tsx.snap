// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductComponent renders product title, image, description, variant selector and price 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gEvEer fKUqjm"
      >
        <div
          style="margin-bottom: 10px;"
        >
          <img
            class="sc-eqUAAy hdyIrO"
            loading="lazy"
            src="https://example.com/image.jpg"
          />
          <div
            class="sc-kAyceB ctPBcu"
          />
          <div
            class="sc-iGgWBj bJOxaZ"
          >
            Product Title
          </div>
          <div>
            <label
              for="selectOption"
              style="display: block;"
            >
              Select a variant
            </label>
            <select
              class="sc-aXZVg bcxOaR"
              id="selectOption"
            >
              <option
                style="display: none;"
                value=""
              >
                Variant 1
              </option>
              <option
                value="Variant 1"
              >
                Variant 1
              </option>
              <option
                value="Variant 2"
              >
                Variant 2
              </option>
            </select>
          </div>
        </div>
        <div
          class="sc-fqkvVR gGlEgu"
        >
          <div
            class="sc-dcJsrY gzIXVR"
          >
            Product Description
          </div>
          <div
            class="sc-imWYAI iLrRnI"
          >
            <button
              class="sc-gsFSXq bNPfJT"
            >
               Buy $100
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gEvEer fKUqjm"
    >
      <div
        style="margin-bottom: 10px;"
      >
        <img
          class="sc-eqUAAy hdyIrO"
          loading="lazy"
          src="https://example.com/image.jpg"
        />
        <div
          class="sc-kAyceB ctPBcu"
        />
        <div
          class="sc-iGgWBj bJOxaZ"
        >
          Product Title
        </div>
        <div>
          <label
            for="selectOption"
            style="display: block;"
          >
            Select a variant
          </label>
          <select
            class="sc-aXZVg bcxOaR"
            id="selectOption"
          >
            <option
              style="display: none;"
              value=""
            >
              Variant 1
            </option>
            <option
              value="Variant 1"
            >
              Variant 1
            </option>
            <option
              value="Variant 2"
            >
              Variant 2
            </option>
          </select>
        </div>
      </div>
      <div
        class="sc-fqkvVR gGlEgu"
      >
        <div
          class="sc-dcJsrY gzIXVR"
        >
          Product Description
        </div>
        <div
          class="sc-imWYAI iLrRnI"
        >
          <button
            class="sc-gsFSXq bNPfJT"
          >
             Buy $100
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
